keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,634 @@
+      - For all the `Secret`s referenced by the resource:
+        - Update the `Secret`'s `Liens` field to remove `kubernetes.io/secret-protection ${version}/${kind}/${namespace}/${name}`.
+  - The controller periodically check `Pod`, `PersistentVolume`, and `VolumeSnapshotContent`
+    - For each resource:
+      - For all the `Secret`s referenced by the resource:
+        - If the `Secret` doesn't have `kubernetes.io/enable-secret-protection: "yes"` annotation, do nothing
+        - Update the `Secret`'s `Liens` field to add `kubernetes.io/secret-protection ${version}/${kind}/${namespace}/${name}`.
+  - The controller preiodically check `Secret`
+    - For each `Secret`:
+      - If the `Secret` doesn't have `kubernetes.io/enable-secret-protection: "yes"` annotation
+        - Remove all the `Liens` that is prefixed with `kubernetes.io/secret-protection`
+      - otherwise:
+        - For all the liens that are prefixed with `kubernetes.io/secret-protection`:
+        - Check if the resource with `${version}/${kind}/${namespace}/${name}` exists, and if it no longer exists, delete the lien.
+
+- ConfigMap protection controller:
+  - The controller watches `Pod`
+    - Create/Update event:
+      - For all the `ConfigMap`s referenced by the `Pod`:
+        - If the `ConfigMap` doesn't have `kubernetes.io/enable-configmap-protection: "yes"` annotation, do nothing
+        - Update the `ConfigMap`'s `Liens` field to add `kubernetes.io/configmap-protection ${version}/${kind}/${namespace}/${name}`.
+    - Delete event:
+      - For all the `ConfigMap`s referenced by the `Pod`:
+        - Update the `ConfigMap`'s `Liens` field to remove `kubernetes.io/configmap-protection ${version}/${kind}/${namespace}/${name}`.
+  - The controller periodically check `Pod`
+    - For each `Pod`:
+      - For all the `ConfigMap`s referenced by the `Pod`:
+        - If the `ConfigMap` doesn't have `kubernetes.io/enable-configmap-protection: "yes"` annotation, do nothing
+        - Update the `ConfigMap`'s `Liens` field to add `kubernetes.io/configmap-protection ${version}/${kind}/${namespace}/${name}`.
+  - The controller preiodically check the `ConfigMap`s
+    - For each `ConfigMap`:
+      - If the `ConfigMap` doesn't have `kubernetes.io/enable-configmap-protection: "yes"` annotation
+        - Remove all the `Liens` that is prefixed with `kubernetes.io/configmap-protection`
+      - otherwise:
+        - For all the liens that are prefixed with `kubernetes.io/configmap-protection`:
+        - Check if the resource with `${version}/${kind}/${namespace}/${name}` exists, and if it no longer exists, delete the lien.
+
+Note that
+  - Periodical checks for referencing resources are needed for handling a case that the annotation is added after referecing resources are created,
+  - Periodical checks for `Secret` or `Configmap` are needed for handling a case that update for a referencing resource is done in a way that remove the reference to `Secret` or `Configmap`,
```

Comment:
How do you periodically check secrets? Are you watching all of them?
If so - then we don't need periodic checks, because every necessary operation will be triggerred by some object change [we will just need some index from secrets/configmaps -> objects referencing it]

Reply:
> How do you periodically check secrets? Are you watching all of them?

I was thinking about polling regularly to call list API and loop over the list
(If we use some frameworks, like kube-controller-manager, it may be achieved just by watching the resource. I understand that some of the frameworks put a resource to a queue even if it doesn't have changes in the resource).

> If so - then we don't need periodic checks, because every necessary operation will be triggerred by some object change [we will just need some index from secrets/configmaps -> objects referencing it]

My intention of these periodic checks is to handle cases that might not perfectly be handled in the actions in watch events.

As for the second bullet case specifically, it is intended to care a below case.

1. A user creates Pod A that uses Secret X and Secret Y 
2. Secret-protection-controller detected creation of Pod A and added lien to Secret X and Secret Y
3. A user updates Pod A to only use Secret X
4. Secret-protection-controller detected update of Pod A and confirmed that Secret X already has its lien

As a result, lien in Secret Y is not removed, properly.

We may find a way to check the difference of the Spec in update events to properly delete the lien on update events. However, update events may not be detected properly, for example due to the controller failure when the update API is called. Therefore, using periodic checks will be still needed, even if we properly handle update events with detecting the difference.